core/state_transition.go

```diff
@@ -357,6 +359,14 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 	// 6. there is no overflow when calculating intrinsic gas
 	// 7. caller has enough balance to cover asset transfer for **topmost** call
 
+	// Clause 0
+	if st.msg.EthCompatible() && !st.evm.ChainConfig().IsDonut(st.evm.BlockNumber) {
+		return nil, ErrEthCompatibleTransactionsNotSupported
+	}
+	if st.msg.EthCompatible() && !(st.msg.FeeCurrency() == nil && st.msg.GatewayFeeRecipient() == nil && st.msg.GatewayFee().Sign() == 0) {
```

Comment:
i'd the `st.msg.FeeCurrency() == nil && st.msg.GatewayFeeRecipient() == nil && st.msg.GatewayFee().Sign() == 0` to a method inside `msg` since they are instrinsic to the `msg` and not to `TransitionDb`. 

Maybe a `Valid() bool` msg or `Validate() error`

Reply:
I removed some of the duplication and moved the logic to methods or stand-alone functions.  This meant moving the error from `core` to `types` to avoid circular input or duplicating the error definition.